plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -136,50 +158,105 @@ class bp_connection_manager {
-            initial_gossip_msg_factory.set_initial_send_buffer(signed_empty);
-            // update gossip_bps
-            auto& prod_idx = gossip_bps.index.get<by_producer>();
-            gossip_bp_peers_message::bp_peer peer{.producer_name = e, .server_address = server_address};
-            peer.sig = self()->sign_compact(*peer_info->key, peer.digest());
-            EOS_ASSERT(peer.sig != signature_type{}, chain::plugin_config_exception,
-                       "Unable to sign bp peer ${p}, private key not found for ${k}", ("p", peer.producer_name)("k", peer_info->key->to_string({})));
-            if (auto i = prod_idx.find(boost::make_tuple(e, boost::cref(server_address))); i != prod_idx.end()) {
-               gossip_bps.index.modify(i, [&peer](auto& v) {
-                  v.sig = peer.sig;
-               });
-            } else {
-               gossip_bps.index.emplace(peer);
+            for (const auto& le : my_bp_account.second) {
+               fc_dlog(self()->get_logger(), "Updating BP gossip_bp_peers_message for ${a} address ${s}", ("a", bp_account)("s", le.server_endpoint));
+                  if (!initial_updated) {
+                     // update initial so always an active one
+                     gossip_bp_peers_message::signed_bp_peer signed_empty{{.producer_name = bp_account}}; // .server_endpoint not set for initial message
+                     signed_empty.sig = self()->sign_compact(*peer_info->key, signed_empty.digest());
+                     EOS_ASSERT(signed_empty.sig != signature_type{}, chain::plugin_config_exception,
+                                "Unable to sign empty gossip bp peer of ${a}, private key not found for ${k}", ("a", bp_account)("k", peer_info->key->to_string({})));
+                     initial_gossip_msg_factory.set_initial_send_buffer(signed_empty);
+                     initial_updated = true;
+                  }
+                  // update gossip_bps
+                  auto& prod_idx = gossip_bps.index.get<by_producer>();
+                  gossip_bp_peers_message::signed_bp_peer peer{
+                     { .producer_name = bp_account,
+                       .server_endpoint = le.server_endpoint,
+                       .outbound_ip_address = le.outbound_ip_address,
+                       .expiration = expire }
+                  };
+                  peer.sig = self()->sign_compact(*peer_info->key, peer.digest());
+                  EOS_ASSERT(peer.sig != signature_type{}, chain::plugin_config_exception,
+                             "Unable to sign bp peer ${p}, private key not found for ${k}", ("p", peer.producer_name)("k", peer_info->key->to_string({})));
+                  if (auto i = prod_idx.find(boost::make_tuple(bp_account, boost::cref(le.server_endpoint))); i != prod_idx.end()) {
+                     gossip_bps.index.modify(i, [&peer](auto& v) {
+                        v.outbound_ip_address = peer.outbound_ip_address;
```

Comment:
Interestingly, this means that we support different `server_endpoint` having the same `outbound_ip_address`, but not different `outbound_ip_address` having the same `server_endpoint`, so for example if we have:

```
p2p-bp-gossip-endpoint = defproda,p2p.proddomain.com:9876,173.194.219.100
p2p-bp-gossip-endpoint = defproda,p2p.proddomain.com:9875,173.194.219.99
```

the second entry will replace the first one in `gossip_bps`. I don't know if it is an issue though.

Reply:
I think we should verify that and not allow it.

Reply:
https://github.com/AntelopeIO/spring/pull/1503/commits/07751bc9849efad8ac40e59ec425c37826477f71